explorer/ast/declaration.h

```diff
@@ -126,6 +126,59 @@ class Declaration : public AstNode {
   bool is_type_checked_ = false;
 };
 
+// A name being declared in a named declaration.
+class DeclaredName {
+ public:
+  struct NameComponent {
+    SourceLocation source_loc;
+    std::string name;
+  };
+
+  // This is required by bison, and should not be used elsewhere.
+  explicit DeclaredName() {}
```

Comment:
I made this explicit because I want to discourage accidental use of this constructor. An explicit default constructor is good enough for bison to use internally, which is the only intended use of this constructor per the comment.